File: Find All Anagrams in a String/Go/solution1.go
package main

func findAnagrams(s string, p string) []int {
	// if p is larger than s there are no matches
	if len(p) > len(s) {
		return []int{}
	}
	// create 2 arrays to store frequnecys. there are 26 letters making the array length 26
	sArr := make([]int, 26)
	pArr := make([]int, 26)
	// create an array to store the start index of anagrams
	startIdxs := make([]int, 0)
	// populate the two arrays with the letter frequences from 0 to length p
	pArr = populateArray(pArr, p, len(p))
	sArr = populateArray(sArr, s, len(p))

	// set the bounds of the window
	// high is exclusive
	high := len(p)
	// low is inclusive
	low := 0

	// loop until high is out of bounds of s
	for high <= len(s) {
		// check if the arrays are equal.
		// if they are add the inclusive low index to the result
		if isEqual(sArr, pArr) {
			startIdxs = append(startIdxs, low)
		}
		// if high is out of bounds break
		if high >= len(s) {
			break
		}
		// decrement low as it wont be in the window
		sArr[s[low]-'a']--
		// increment high as it will not be in the window
		sArr[s[high]-'a']++
		// increment both low and high values
		low++
		high++
	}

	return startIdxs
}

// populateArray populates the array witht eh freqeucy of letters from 0 to end of str. Returns the updated array
func populateArray(arr []int, str string, end int) []int {
	for i := 0; i < end; i++ {
		arr[str[i]-'a']++
	}

	return arr
}

// isEquals compares the values in to arrays and returns false if they are not equal
func isEqual(arr1, arr2 []int) bool {
	for i, _ := range arr1 {
		if arr1[i] != arr2[i] {
			return false
		}
	}

	return true
}

File: Find All Anagrams in a String/JavaScript/solution1.js
/**
 * @param {string} s
 * @param {string} p
 * @return {number[]}
 */
 var findAnagrams = function(s, p) {
    // load a map of the first p.length of s
    let sMap = loadMap({}, s, p.length)
    // load a map of the first p length of p
    let pMap = loadMap({}, p, p.length)
    let pLength = p.length
    // the bounds of the window. low is inclusive high is exclusive
    let low = 0;
    let high = pLength;
    // the return value of the function
    let anagramsStart = []
    // loop until high is equal to s length
    while(high <= s.length){
        // check if the map are equal
        if(isEqual(sMap, pMap)) {
            anagramsStart.push(low)
        }

        // update the maps for the new window
        // Decrement the inclusive low value
        sMap[s[low]]--
        // increment the exclusive high value
        if(sMap[s[high]]) {
            sMap[s[high]]++
        } else {
            sMap[s[high]] = 1
        }
            // move the window to the right
        high++
        low++
    }
    
    return anagramsStart
    
};

// isEqual compares maps. Returns false if key/value are nto equal
var isEqual = function(sMap, pMap) {
    for(key in pMap) {
        if(sMap[key] != pMap[key]) {
            return false
        }
    }
    return true
}

// loadMap populates the map with the first len frequencies 
var loadMap = function(charMap, str, len) {
    for(let i = 0; i < len; i++){
        let char = str[i]
        if(charMap[char]) {
            charMap[char]++
        } else {
            charMap[char] = 1
        }
    }
    return charMap
}